util/comparator.cc

```diff
@@ -216,6 +217,63 @@ class ReverseBytewiseComparatorImpl : public BytewiseComparatorImpl {
     return -a.compare(b);
   }
 };
+
+/* Comparator with 64-bit integer timestamp.
+This is still experimental, use it with caution.
+*/
+class ComparatorWithU64TsImpl : public Comparator {
```

Comment:
Since this comparator is now in ROCKSDB_NAMESPACE and will be exposed via public API, we need improve it more.
Currently, the constructor hard-codes `BytewiseComparator()`, which is not necessary. We should be able to wrap an arbitrary comparator (without ts) using this class. When wrapping around a pointer to another comparator, we should consider object lifetime.

Reply:
Its slightly tricky as I think about it.
If we can pass any arbitrarory compare, we need a way to store/retrieve it from the sst file properties

Reply:
Maybe there is some confusion here.
Say we want to support `BytewiseComparatorWithU64Ts`, `ReverseBytewiseComparatorWithU64Ts` and `FancyComparatorWithU64Ts()`. In the object registry, we have the following:
```
"leveldb.BytewiseComparator" => BytewiseComparator()
"rocksdb.ReverseBytewiseComparator" => ReverseBytewiseComparator()
"rocksdb.BytewiseComparator.u64ts" => BytewiseComparatorWithU64Ts() (this does not have to be API yet)
"rocksdb.ReverseBytewiseComparator.u64ts"=> ReverseBytewiseComparatorWithU64Ts() (does not have to be public API yet)
"rocksdb.FancyComparator.u64ts" => FancyComparatorWithU64Ts() // I made this up
```

`BytewiseComparatorWithU64Ts() `, `ReverseBytewiseComparatorWithU64Ts()` and `FancyComparatorWithU64Ts()` can all use `ComparatorWithU64TsImpl` class. In each of them, they can determine how they construct `ComparatorWithU64TsImpl` object.

Reply:
Updated this based on the internal chat